### DISCUSSION TITLE
Getting the Id of the widget sending the message

### DISCUSSION THREAD

**Discussion by duncanf95** [hecrj/iced#1771] [Q&A] [answered]
hey I'm trying to create an app where there would be multiple similar rows for data input (line items as I would refer to them)

so I want to have rows, each row has a text input, each text input should be exactly the same aside from their ID. So each text input calls the same message for onChange event.

I can't just add more messages as the number of rows is defined by the user during the running of the application, and this would also be quite inefficient.

So how do I determine which text input has sent the message? I'm sorry if this has a really obvious answer but I couldn't find a good example for this. I've left a screenshot for more context.

![Screenshot 2023-03-27 at 14 46 07](https://user-images.githubusercontent.com/22371732/227959529-9482052a-3b37-45d0-a3a6-9cff882135d8.png)

**Comment by hecrj** [comment, marked as answer]
You could add the row index to the message to identify the input producing the messages:

```rust
enum Message {
    // ...
    DataChanged(usize, String),
    TagChanged(usize, String),
    // ...
}
```

**Reply by duncanf95** [reply]
ok sorry for delay, have tried this but when I provide the `Message::DataChanged` option to the text input it says its expecting a function input with a single argument, the ID is being set on the text input. How would I tell the text input to accept an enum with more than one argument?

**Reply by hecrj** [reply]
A `TextInput` expects a closure that produces a `Message` given a `String`. You can capture outside variables in this closure.

```rust
let inputs = (0..n_rows).map(|row| text_input(/*... */, move |new_value| Message::DataChanged(row, new_value)));
```

**Reply by hecrj** [reply]
The `todos` example does this:

https://github.com/iced-rs/iced/blob/4e409bb383be103a6ecd66d1e52a0a8d41fd7ddb/examples/todos/src/main.rs#L221-L244
